<a href="https://colab.research.google.com/github/Rivaldop/metodologidatascience/blob/main/Data_Preparation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### DATA PREPARATION

Unduh Data : https://github.com/realpython/python-data-cleaning/blob/master/Datasets/BL-Flickr-Images-Book.csv

1. import library yang dibutuhkan 

In [3]:
import pandas as pd


In [4]:
import numpy as np

In [11]:
df = pd.read_csv('BL-Flickr-Images-Book.csv')
data.head()

Identifier             Edition Statement      Place of Publication  \
0         206                           NaN                    London   
1         216                           NaN  London; Virtue & Yorston   
2         218                           NaN                    London   
3         472                           NaN                    London   
4         480  A new edition, revised, etc.                    London   

  Date of Publication              Publisher  \
0         1879 [1878]       S. Tinsley & Co.   
1                1868           Virtue & Co.   
2                1869  Bradbury, Evans & Co.   
3                1851          James Darling   
4                1857   Wertheim & Macintosh   

                                               Title     Author  \
0                  Walter Forbes. [A novel.] By A. A      A. A.   
1  All for Greed. [A novel. The dedication signed...  A., A. A.   
2  Love the Avenger. By the author of “All for Gr...  A., A. A.   
3  Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
4  [The World in which I live, and my place in it...  A., E. S.   

                                   Contributors  Corporate Author  \
0                               FORBES, Walter.               NaN   
1  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
2  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
3                   Appleyard, Ernest Silvanus.               NaN   
4                           BROOME, John Henry.               NaN   

   Corporate Contributors Former owner  Engraver Issuance type  \
0                     NaN          NaN       NaN   monographic   
1                     NaN          NaN       NaN   monographic   
2                     NaN          NaN       NaN   monographic   
3                     NaN          NaN       NaN   monographic   
4                     NaN          NaN       NaN   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   
2  http://www.flickr.com/photos/britishlibrary/ta...   
3  http://www.flickr.com/photos/britishlibrary/ta...   
4  http://www.flickr.com/photos/britishlibrary/ta...   

                            Shelfmarks  
0    British Library HMNTS 12641.b.30.  
1    British Library HMNTS 12626.cc.2.  
2    British Library HMNTS 12625.dd.1.  
3  British Library HMNTS 10369.bbb.15.  
4     British Library HMNTS 9007.d.28.

In [15]:
yg_dihapus = ['Edition Statement','Corporate Author']

In [16]:
df.drop(yg_dihapus, inplace=True, axis=1)

In [18]:
df.head()

Identifier      Place of Publication Date of Publication  \
0         206                    London         1879 [1878]   
1         216  London; Virtue & Yorston                1868   
2         218                    London                1869   
3         472                    London                1851   
4         480                    London                1857   

               Publisher                                              Title  \
0       S. Tinsley & Co.                  Walter Forbes. [A novel.] By A. A   
1           Virtue & Co.  All for Greed. [A novel. The dedication signed...   
2  Bradbury, Evans & Co.  Love the Avenger. By the author of “All for Gr...   
3          James Darling  Welsh Sketches, chiefly ecclesiastical, to the...   
4   Wertheim & Macintosh  [The World in which I live, and my place in it...   

      Author                                  Contributors  \
0      A. A.                               FORBES, Walter.   
1  A., A. A.  BLAZE DE BURY, Marie Pauline Rose - Baroness   
2  A., A. A.  BLAZE DE BURY, Marie Pauline Rose - Baroness   
3  A., E. S.                   Appleyard, Ernest Silvanus.   
4  A., E. S.                           BROOME, John Henry.   

   Corporate Contributors Former owner  Engraver Issuance type  \
0                     NaN          NaN       NaN   monographic   
1                     NaN          NaN       NaN   monographic   
2                     NaN          NaN       NaN   monographic   
3                     NaN          NaN       NaN   monographic   
4                     NaN          NaN       NaN   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   
2  http://www.flickr.com/photos/britishlibrary/ta...   
3  http://www.flickr.com/photos/britishlibrary/ta...   
4  http://www.flickr.com/photos/britishlibrary/ta...   

                            Shelfmarks  
0    British Library HMNTS 12641.b.30.  
1    British Library HMNTS 12626.cc.2.  
2    British Library HMNTS 12625.dd.1.  
3  British Library HMNTS 10369.bbb.15.  
4     British Library HMNTS 9007.d.28.

In [20]:
data['Identifier'].is_unique

True

In [21]:
df = df.set_index('Identifier')

In [23]:
df.head()

Place of Publication Date of Publication  \
Identifier                                                 
206                           London         1879 [1878]   
216         London; Virtue & Yorston                1868   
218                           London                1869   
472                           London                1851   
480                           London                1857   

                        Publisher  \
Identifier                          
206              S. Tinsley & Co.   
216                  Virtue & Co.   
218         Bradbury, Evans & Co.   
472                 James Darling   
480          Wertheim & Macintosh   

                                                        Title     Author  \
Identifier                                                                 
206                         Walter Forbes. [A novel.] By A. A      A. A.   
216         All for Greed. [A novel. The dedication signed...  A., A. A.   
218         Love the Avenger. By the author of “All for Gr...  A., A. A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
480         [The World in which I live, and my place in it...  A., E. S.   

                                            Contributors  \
Identifier                                                 
206                                      FORBES, Walter.   
216         BLAZE DE BURY, Marie Pauline Rose - Baroness   
218         BLAZE DE BURY, Marie Pauline Rose - Baroness   
472                          Appleyard, Ernest Silvanus.   
480                                  BROOME, John Henry.   

            Corporate Contributors Former owner  Engraver Issuance type  \
Identifier                                                                
206                            NaN          NaN       NaN   monographic   
216                            NaN          NaN       NaN   monographic   
218                            NaN          NaN       NaN   monographic   
472                            NaN          NaN       NaN   monographic   
480                            NaN          NaN       NaN   monographic   

                                                   Flickr URL  \
Identifier                                                      
206         http://www.flickr.com/photos/britishlibrary/ta...   
216         http://www.flickr.com/photos/britishlibrary/ta...   
218         http://www.flickr.com/photos/britishlibrary/ta...   
472         http://www.flickr.com/photos/britishlibrary/ta...   
480         http://www.flickr.com/photos/britishlibrary/ta...   

                                     Shelfmarks  
Identifier                                       
206           British Library HMNTS 12641.b.30.  
216           British Library HMNTS 12626.cc.2.  
218           British Library HMNTS 12625.dd.1.  
472         British Library HMNTS 10369.bbb.15.  
480            British Library HMNTS 9007.d.28.

In [22]:
df.loc[472]

Place of Publication                                                 London
Date of Publication                                                    1851
Publisher                                                     James Darling
Title                     Welsh Sketches, chiefly ecclesiastical, to the...
Author                                                            A., E. S.
Contributors                                    Appleyard, Ernest Silvanus.
Corporate Contributors                                                  NaN
Former owner                                                            NaN
Engraver                                                                NaN
Issuance type                                                   monographic
Flickr URL                http://www.flickr.com/photos/britishlibrary/ta...
Shelfmarks                              British Library HMNTS 10369.bbb.15.
Name: 472, dtype: object

3. merapikan fields

In [24]:
df.dtypes.value_counts()

object     10
float64     2
dtype: int64

In [25]:
df.loc[1905:,'Date of Publication'].head(15)

Identifier
1905                  1888
1929           1839, 38-54
2836                  1897
2854                  1865
2956               1860-63
2957                  1873
3017                  1866
3131                  1899
4598                  1814
4884                  1820
4976                  1800
5382    1847, 48 [1846-48]
5385               [1897?]
5389               [1897?]
5432                  1893
Name: Date of Publication, dtype: object

~ hilangkan tgl lain dalam kurung siku
~ hilangkan rentang tanggal
~ hilangkan tanggal yang gajelas[1897?]--> NaN
~ konversi NaN

In [30]:
regex = r'^(\d{4})'

In [31]:
ekstrak = df['Date of Publication'].str.extract(r'^(\d{4})',expand=False)

In [32]:
df.loc[667]

Place of Publication                    pp. 40. G. Bryan & Co: Oxford, 1898
Date of Publication                                                     NaN
Publisher                                                               NaN
Title                     The Coming of Spring, and other poems. By J. A...
Author                                                        A., J.|A., J.
Contributors                                  ANDREWS, J. - Writer of Verse
Corporate Contributors                                                  NaN
Former owner                                                            NaN
Engraver                                                                NaN
Issuance type                                                   monographic
Flickr URL                http://www.flickr.com/photos/britishlibrary/ta...
Shelfmarks                               British Library HMNTS 011652.g.73.
Name: 667, dtype: object

In [33]:
df.loc[4157862]

Place of Publication                                    Newcastle-upon-Tyne
Date of Publication                                                    1867
Publisher                                                        T. Fordyce
Title                     Local Records; or, Historical Register of rema...
Author                        FORDYCE, T. - Printer, of Newcastle-upon-Tyne
Contributors               SYKES, John - Bookseller, of Newcastle-upon-Tyne
Corporate Contributors                                                  NaN
Former owner                                                            NaN
Engraver                                                                NaN
Issuance type                                                   monographic
Flickr URL                http://www.flickr.com/photos/britishlibrary/ta...
Shelfmarks                British Library HMNTS|British Library HMNTS 01...
Name: 4157862, dtype: object

In [34]:
df['Place of Publication'].tail(15)

Identifier
4115138    G. Eld: London, 1608
4116063                  London
4117526                 Leipzig
4117583                  London
4117749                New York
4117751                New York
4117752                  London
4156359                New York
4157746                  London
4157862     Newcastle-upon-Tyne
4158088                  London
4158128                   Derby
4159563                  London
4159587     Newcastle upon Tyne
4160339                  London
Name: Place of Publication, dtype: object

In [35]:
df.loc[4115138]

Place of Publication                                   G. Eld: London, 1608
Date of Publication                                                     NaN
Publisher                                                               NaN
Title                     The Revenger's Tagaedie. As it hath beene sund...
Author                                                                  NaN
Contributors                                               TOURNEUR, Cyril.
Corporate Contributors                                                  NaN
Former owner                                                            NaN
Engraver                                                                NaN
Issuance type                                                   monographic
Flickr URL                http://www.flickr.com/photos/britishlibrary/ta...
Shelfmarks                British Library HMNTS 644.c.80.|British Librar...
Name: 4115138, dtype: object

In [40]:
publikasi = df['Place of Publication']
london = publikasi.str.contains('London')
london[:5]

Identifier
206    True
216    True
218    True
472    True
480    True
Name: Place of Publication, dtype: bool

In [42]:
oxford = publikasi.str.contains ('Oxford')

In [49]:
df['Place of Publication'] = np.where(london,'London', 
                                      np.where(oxford, 'Oxford',
                                      publikasi.str.replace ('-', ' ')))

In [51]:
df['Place of Publication'].head(15)

Identifier
206          London
216          London
218          London
472          London
480          London
481          London
519          London
667          Oxford
874          London
1143         London
1280       Coventry
1808    Christiania
1905        Firenze
1929      Amsterdam
2836         Savona
Name: Place of Publication, dtype: object

####DATA SET BARU 

5. membersihkan dataset dengan applymap

In [39]:
university_town = []
with open("university_towns.txt") as file:
  for line in file:
    if '[edit]'in line:
      state = line 
    else:
        university_town.append((state,line))
    university_town[:5]

In [ ]:
df_kota = pd.DataFrame(university_town, columns=['State','RegionName'])
df_kota.head(15)

State                                         RegionName
0    Alabama[edit]\n                    Auburn (Auburn University)[1]\n
1    Alabama[edit]\n           Florence (University of North Alabama)\n
2    Alabama[edit]\n  Jacksonville (Jacksonville State University)[2]\n
3    Alabama[edit]\n       Livingston (University of West Alabama)[2]\n
4    Alabama[edit]\n         Montevallo (University of Montevallo)[2]\n
5    Alabama[edit]\n                        Troy (Troy University)[2]\n
6    Alabama[edit]\n  Tuscaloosa (University of Alabama, Stillman Co...
7    Alabama[edit]\n                Tuskegee (Tuskegee University)[5]\n
8     Alaska[edit]\n    Fairbanks (University of Alaska Fairbanks)[2]\n
9    Arizona[edit]\n       Flagstaff (Northern Arizona University)[6]\n
10   Arizona[edit]\n                 Tempe (Arizona State University)\n
11   Arizona[edit]\n                   Tucson (University of Arizona)\n
12  Arkansas[edit]\n  Arkadelphia (Henderson State University, Ouach...
13  Arkansas[edit]\n  Conway (Central Baptist College, Hendrix Colle...
14  Arkansas[edit]\n         Fayetteville (University of Arkansas)[7]\n

In [ ]:
def get_citystate(item):
  if '(' in item:
      return item[:item.find('(')]
  elif '[' in item:
    return item[:item.find('[')]
  else:
      return item


In [ ]:
df_kota = df_kota.applymap(get_citystate)

In [ ]:
df_kota.head

<bound method NDFrame.head of          State      RegionName
0      Alabama         Auburn 
1      Alabama       Florence 
2      Alabama   Jacksonville 
3      Alabama     Livingston 
4      Alabama     Montevallo 
..         ...             ...
512  Wisconsin    River Falls 
513  Wisconsin  Stevens Point 
514  Wisconsin       Waukesha 
515  Wisconsin     Whitewater 
516    Wyoming        Laramie 

[517 rows x 2 columns]>